From: CloudFox-Kernel <kernel@cloudfox.dev>
Date: Mon, 3 Aug 2026 00:00:00 +0000
Subject: [PATCH 3/3] le9uo workingset protection: anon_min_ratio,
 clean_low_ratio, clean_min_ratio

This patch adds the le9uo workingset protection mechanism, which provides
fine-grained control over memory reclaim behavior:

1. sysctl workingset_protection (on/off) - master switch for the feature
2. sysctl anon_min_ratio - hard protection of anonymous pages (default 15%)
3. sysctl clean_low_ratio - best-effort protection of clean file pages (default 0%)
4. sysctl clean_min_ratio - hard protection of clean file pages (default 15%)

When anon_min_ratio is set, anonymous pages below that threshold won't be
reclaimed under any conditions. This prevents excessive swap thrashing
when memory is overfilled by compressed data (e.g., ZRAM).

When clean_low/clean_min_ratio is set, the reclaim logic will favor
scanning anonymous pages over clean file pages when the amount of clean
file pages falls below the threshold.

Signed-off-by: CloudFox-Kernel <kernel@cloudfox.dev>
---
 mm/Kconfig        |  66 +++++++++++++++++++
 include/linux/mm.h |   8 +++
 mm/vmscan.c       | 115 ++++++++++++++++++++++++++++++++
 kernel/sysctl.c   |  48 +++++++++++++
 4 files changed, 237 insertions(+)

diff --git a/mm/Kconfig b/mm/Kconfig
index 6c21c92..8a5e6b0 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -1272,6 +1272,72 @@ config COMPAT_BRK
 
 	  On non-ancient systems, it is safe to say N.
 
+config ANON_MIN_RATIO
+	int "Default value for vm.anon_min_ratio"
+	depends on SYSCTL
+	range 0 100
+	default 15
+	help
+	  This option sets the default value for vm.anon_min_ratio sysctl knob.
+
+	  The vm.anon_min_ratio sysctl knob provides *hard* protection of
+	  anonymous pages. The anonymous pages on the current node won't be
+	  reclaimed under any conditions when their amount is below
+	  vm.anon_min_ratio. This knob may be used to prevent excessive swap
+	  thrashing when anonymous memory is low (for example, when memory is
+	  going to be overfilled by compressed data of zram module).
+
+	  Setting this value too high (close to 100) can result in
+	  inability to swap and can lead to early OOM under memory pressure.
+
+config CLEAN_LOW_RATIO
+	int "Default value for vm.clean_low_ratio"
+	depends on SYSCTL
+	range 0 100
+	default 0
+	help
+	  This option sets the default value for vm.clean_low_ratio sysctl knob.
+
+	  The vm.clean_low_ratio sysctl knob provides *best-effort*
+	  protection of clean file pages. When the amount of clean file pages
+	  on the current node is below vm.clean_low_ratio, the reclaim logic
+	  will be more reluctant to evict clean file pages, trying to swap
+	  anonymous pages first.
+
+	  Setting it to a high value may result in a early eviction of
+	  anonymous pages into the swap space by attempting to hold the
+	  protected amount of clean file pages in memory.
+
+config CLEAN_MIN_RATIO
+	int "Default value for vm.clean_min_ratio"
+	depends on SYSCTL
+	range 0 100
+	default 15
+	help
+	  This option sets the default value for vm.clean_min_ratio sysctl knob.
+
+	  The vm.clean_min_ratio sysctl knob provides *hard* protection of
+	  clean file pages. When the amount of clean file pages on the current
+	  node is below vm.clean_min_ratio, they won't be reclaimed under any
+	  conditions.
+
+	  Setting it to a high value may result in inability to reclaim file
+	  pages and can lead to early OOM under memory pressure.
+
 config MMU
 	bool
 	default y
diff --git a/include/linux/mm.h b/include/linux/mm.h
index 4e6b2a0..7e6a3b0 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -214,6 +214,14 @@ extern int sysctl_max_map_count;
 
 extern unsigned long sysctl_user_reserve_kbytes;
 extern unsigned long sysctl_admin_reserve_kbytes;
+
+extern int sysctl_workingset_protection;
+extern u8 sysctl_anon_min_ratio;
+extern u8 sysctl_clean_low_ratio;
+extern u8 sysctl_clean_min_ratio;
+int vm_workingset_protection_update_handler(struct ctl_table *table, int write,
+					     void *buffer, size_t *lenp,
+					     loff_t *ppos);
 #endif /* CONFIG_SYSCTL */
 
 /* Architectures with odd memory protection constraints */
diff --git a/mm/vmscan.c b/mm/vmscan.c
index 6d5c78e..9a1b2c3 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -132,6 +132,15 @@ struct scan_control {
 	/* The file pages on the current node are dangerously low */
 	unsigned int file_is_tiny:1;
 
+	/* The anonymous pages on the current node are below vm.anon_min_ratio */
+	unsigned int anon_below_min:1;
+
+	/* The clean file pages on the current node are below vm.clean_low_ratio */
+	unsigned int clean_below_low:1;
+
+	/* The clean file pages on the current node are below vm.clean_min_ratio */
+	unsigned int clean_below_min:1;
+
 	/* Allocation order */
 	s8 order;
 
@@ -192,6 +201,16 @@ int vm_swappiness = 60;
 
 int sysctl_workingset_protection __read_mostly = 1;
+u8 sysctl_anon_min_ratio  __read_mostly = CONFIG_ANON_MIN_RATIO;
+u8 sysctl_clean_low_ratio __read_mostly = CONFIG_CLEAN_LOW_RATIO;
+u8 sysctl_clean_min_ratio __read_mostly = CONFIG_CLEAN_MIN_RATIO;
+static u64 sysctl_anon_min_ratio_kb  __read_mostly;
+static u64 sysctl_clean_low_ratio_kb __read_mostly;
+static u64 sysctl_clean_min_ratio_kb __read_mostly;
+static u64 workingset_protection_prev_totalram __read_mostly;
+
 #define DEF_KSWAPD_THREADS_PER_NODE 1
 static int kswapd_threads = DEF_KSWAPD_THREADS_PER_NODE;
 static int __init kswapd_per_node_setup(char *str)
@@ -2410,6 +2429,14 @@ static void get_scan_count(struct lruvec *lruvec, struct scan_control *sc,
 	}
 
 	/*
+	 * Force-scan anon if clean file pages is under vm.clean_low_ratio
+	 * or vm.clean_min_ratio.
+	 */
+	if (sc->clean_below_low || sc->clean_below_min) {
+		scan_balance = SCAN_ANON;
+		goto out;
+	}
+
 	/*
 	 * If there is enough inactive page cache, we do not reclaim
 	 * anything from the anonymous working right now. But when balancing
@@ -2567,6 +2594,12 @@ static void get_scan_count(struct lruvec *lruvec, struct scan_control *sc,
 			BUG();
 		}
 
+		/*
+		 * Hard protection of the working set.
+		 */
+		if (file ? sc->clean_below_min : sc->anon_below_min)
+			scan = 0;
+
 		nr[lru] = scan;
 	}
 }
@@ -2586,6 +2619,86 @@ static void shrink_lruvec(struct lruvec *lruvec, struct scan_control *sc)
 	blk_finish_plug(&plug);
 }
 
+int vm_workingset_protection_update_handler(struct ctl_table *table, int write,
+					     void *buffer, size_t *lenp,
+					     loff_t *ppos)
+{
+	int ret = proc_dou8vec_minmax(table, write, buffer, lenp, ppos);
+
+	if (ret || !write)
+		return ret;
+
+	workingset_protection_prev_totalram = 0;
+
+	return 0;
+}
+
+static void prepare_workingset_protection(pg_data_t *pgdat, struct scan_control *sc)
+{
+	unsigned long node_mem_total;
+	struct sysinfo i;
+
+	if (!sysctl_workingset_protection) {
+		sc->anon_below_min = 0;
+		sc->clean_below_low = 0;
+		sc->clean_below_min = 0;
+		return;
+	}
+
+	if (likely(sysctl_anon_min_ratio ||
+		   sysctl_clean_low_ratio ||
+		   sysctl_clean_min_ratio)) {
+#ifdef CONFIG_NUMA
+		si_meminfo_node(&i, pgdat->node_id);
+#else
+		si_meminfo(&i);
+#endif
+		node_mem_total = i.totalram;
+
+		if (unlikely(workingset_protection_prev_totalram != node_mem_total)) {
+			sysctl_anon_min_ratio_kb =
+				node_mem_total * sysctl_anon_min_ratio / 100;
+			sysctl_clean_low_ratio_kb =
+				node_mem_total * sysctl_clean_low_ratio / 100;
+			sysctl_clean_min_ratio_kb =
+				node_mem_total * sysctl_clean_min_ratio / 100;
+			workingset_protection_prev_totalram = node_mem_total;
+		}
+	}
+
+	if (sysctl_anon_min_ratio) {
+		unsigned long reclaimable_anon;
+
+		reclaimable_anon =
+			node_page_state(pgdat, NR_ACTIVE_ANON) +
+			node_page_state(pgdat, NR_INACTIVE_ANON) +
+			node_page_state(pgdat, NR_ISOLATED_ANON);
+
+		sc->anon_below_min = reclaimable_anon < sysctl_anon_min_ratio_kb;
+	} else {
+		sc->anon_below_min = 0;
+	}
+
+	if (sysctl_clean_low_ratio || sysctl_clean_min_ratio) {
+		unsigned long reclaimable_file, dirty, clean;
+
+		reclaimable_file =
+			node_page_state(pgdat, NR_ACTIVE_FILE) +
+			node_page_state(pgdat, NR_INACTIVE_FILE) +
+			node_page_state(pgdat, NR_ISOLATED_FILE);
+		dirty = node_page_state(pgdat, NR_FILE_DIRTY);
+
+		if (likely(reclaimable_file > dirty))
+			clean = reclaimable_file - dirty;
+		else
+			clean = 0;
+
+		sc->clean_below_low = clean < sysctl_clean_low_ratio_kb;
+		sc->clean_below_min = clean < sysctl_clean_min_ratio_kb;
+	} else {
+		sc->clean_below_low = 0;
+		sc->clean_below_min = 0;
+	}
+}
+
 /* Use inline/__always_inline to ensure the compiler won't split the function */
 static inline void pgdat_clear_lruvecs(struct pglist_data *pgdat)
 {
@@ -2910,6 +3023,8 @@ static void shrink_node(pg_data_t *pgdat, struct scan_control *sc)
 
 	target_lruvec = mem_cgroup_lruvec(sc->target_mem_cgroup, pgdat);
 
+	prepare_workingset_protection(pgdat, sc);
+
 again:
 	memset(&sc->nr, 0, sizeof(sc->nr));
 
diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index 5e6c7a9..b8d2e4f 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -3188,6 +3188,54 @@ static struct ctl_table vm_table[] = {
 	},
 #endif
 	{
+		.procname	= "workingset_protection",
+		.data		= &sysctl_workingset_protection,
+		.maxlen		= sizeof(int),
+		.mode		= 0644,
+		.proc_handler	= &proc_dointvec_minmax,
+		.extra1		= SYSCTL_ZERO,
+		.extra2		= SYSCTL_ONE,
+	},
+	{
+		.procname	= "anon_min_ratio",
+		.data		= &sysctl_anon_min_ratio,
+		.maxlen		= sizeof(u8),
+		.mode		= 0644,
+		.proc_handler	= &vm_workingset_protection_update_handler,
+		.extra1		= SYSCTL_ZERO,
+		.extra2		= SYSCTL_ONE_HUNDRED,
+	},
+	{
+		.procname	= "clean_low_ratio",
+		.data		= &sysctl_clean_low_ratio,
+		.maxlen		= sizeof(u8),
+		.mode		= 0644,
+		.proc_handler	= &vm_workingset_protection_update_handler,
+		.extra1		= SYSCTL_ZERO,
+		.extra2		= SYSCTL_ONE_HUNDRED,
+	},
+	{
+		.procname	= "clean_min_ratio",
+		.data		= &sysctl_clean_min_ratio,
+		.maxlen		= sizeof(u8),
+		.mode		= 0644,
+		.proc_handler	= &vm_workingset_protection_update_handler,
+		.extra1		= SYSCTL_ZERO,
+		.extra2		= SYSCTL_ONE_HUNDRED,
+	},
+	{
 		.procname	= "user_reserve_kbytes",
 		.data		= &sysctl_user_reserve_kbytes,
 		.maxlen		= sizeof(sysctl_user_reserve_kbytes),
-- 
2.34.1